Addition of OpenShift provider

abonas opened the pull request:
Contains provider (with a mixin to avoid inheritance from k8s provider), parser, refresher,
refresh worker and persistence of container_project and container_route

abonas commented:
@Fryguy , @blomquisg  the provider is missing a second call to kubeclient to get k8s inventory.
Advice is appreciated how to add that part, since it's not that common in other providers to collect inventory from more than one source.

abonas commented:
@miq-bot add_label providers/containers

Fryguy commented on vmdb/app/helpers/ems_container_helper/textual_summary.rb:
```diff
@@ -10,6 +10,9 @@ def textual_group_properties
 
   def textual_group_relationships
     items = %w(container_nodes container_services container_groups)
+    if @ems.class.ems_type.eql? 'openshift'
```
Prefer `==` over `.eql?`  However, even so, this should just be `@ems.kind_of?(EmsOpenshift)`

abonas replied:
> refer == over .eql? However, even so, this should just be @ems.kind_of?(EmsOpenshift)

sure, will do "kind_of"

Fryguy commented on vmdb/app/helpers/ems_container_helper/textual_summary.rb:
```diff
@@ -10,6 +10,9 @@ def textual_group_properties
   def textual_group_relationships
     items = %w(container_nodes container_services container_groups)
+    if @ems.class.ems_type.eql? 'openshift'
+      items.concat(%w(container_routes container_projects))
```
Actually, even further, at the top level here, you don't even need the conditional.  Instead the individual line item methods below could possible `return nil if @ems.kind_of?(EmsOpenstack)`.  Then you get the logic living with the individual methods.

abonas replied:
> Actually, even further, at the top level here, you don't even need the conditional. Instead the individual line item methods below could possible return nil if @ems.kind_of?(EmsOpenstack). Then you get the logic living with the individual methods.

but then it means repeating this conditional logic multiple times instead of having it clear in one place.

Fryguy commented on vmdb/app/models/ems_kubernetes.rb:
```diff
@@ -1,7 +1,5 @@
-  has_many :container_nodes,                      :foreign_key => :ems_id, :dependent => :destroy
-  has_many :container_groups,                     :foreign_key => :ems_id, :dependent => :destroy
-  has_many :container_services,                   :foreign_key => :ems_id, :dependent => :destroy
+  include ContainerProviderMixin
```
This should work nicely :+1:

Fryguy commented on vmdb/app/models/ems_openshift.rb:
```diff
@@ -0,0 +1,34 @@
+  end
+
+  def self.description
+    @description ||= "Openshift".freeze
```
OpenShift is a capital O and capital S.  I'm ok with the constants like `EmsOpenshift` and `OpenshiftClient` since the underscore/lowercase (i.e. openshift) matches with the upstream naming, and users generaly won't "see" those words.  However any place that's UI facing should say "OpenShift".  To take it a step further, I think the official naming must be one of "Red Hat OpenShift" or "OpenShift by Red Hat" (you'd probably have to ask what the right string is)

abonas replied:
> OpenShift is a capital O and capital S. I'm ok with the constants like EmsOpenshift and OpenshiftClient since the underscore/lowercase (i.e. openshift) matches with the upstream naming, and users generaly won't "see" those words. However any place that's UI facing should say "OpenShift". To take it a step further, I think the official naming must be one of "Red Hat OpenShift" or "OpenShift by Red Hat" (you'd probably have to ask what the right string is)

will rename the description to OpenShift.
the official naming is per how it's named in upstream.

Fryguy commented on vmdb/app/models/ems_openshift.rb:
```diff
@@ -0,0 +1,34 @@
+    # TODO: support real authentication using certificates
+    kube.ssl_options(:verify_ssl => OpenSSL::SSL::VERIFY_NONE)
+    kube
+  end
```
Should this be a class method to match the raw_connect method. Something like `def self.raw_kubernetes_connect` or something.  Then, typically an instance that's service specific is exposed.  See something similar for the EmsAmazon.

Fryguy commented on vmdb/app/models/ems_refresh/parsers/openshift.rb:
```diff
@@ -0,0 +1,43 @@
+      new_result.except!(:namespace)
+      new_result.merge!(
+          :labels       => parse_labels(project),
+          :display_name => project.displayName
```
2 space indent.

Fryguy replied:
And also elsewhere in this file.

Fryguy commented on vmdb/app/models/ems_refresh/parsers/openshift.rb:
```diff
@@ -0,0 +1,43 @@
+          :labels       => parse_labels(project),
+          :display_name => project.displayName
+      )
+      p new_result
```
This is debug and should be removed.

Fryguy commented on vmdb/app/models/ems_refresh/refreshers/openshift_refresher.rb:
```diff
@@ -0,0 +1,14 @@
+module EmsRefresh
+  module Refreshers
+   class OpenshiftRefresher < BaseRefresher
```
Should this be a subclass of KubernetesRefresher since the parser is also a subclass?

abonas replied:
> Should this be a subclass of KubernetesRefresher since the parser is also a subclass?

I don't think so, because the parse_inventory method will look different because of another call to connect as discussed below wrt to amazon use case.

Fryguy commented on vmdb/app/models/ems_refresh/refreshers/openshift_refresher.rb:
```diff
@@ -0,0 +1,14 @@
+    def parse_inventory(ems, _targets = nil)
+      all_entities = ems.with_provider_connection(&:all_entities)
+      # all_entities.merge!(ems.kubernetes_connect.all_entities)
+      EmsRefresh.log_inv_debug_trace(all_entities, "inv_hash:")
```
Does this need an `if DEBUG_TRACE` or does the method itself handle it (I can't remember)

abonas replied:
> Does this need an if DEBUG_TRACE or does the method itself handle it (I can't remember)

that method starts with "return unless debug_trace" so it looks like it handles it itself

Fryguy replied:
Great thanks.

Fryguy commented on vmdb/app/models/ems_refresh/refreshers/openshift_refresher.rb:
```diff
@@ -0,0 +1,14 @@
+      EmsRefresh::Parsers::Openshift.ems_inv_to_hashes(all_entities)
+    end
+   end
+  end
```
There is a one-space alignment issue somewhere.

Fryguy commented on vmdb/app/models/mixins/container_provider_mixin.rb:
```diff
@@ -0,0 +1,42 @@
+    has_many :container_groups,                     :foreign_key => :ems_id, :dependent => :destroy
+    has_many :container_services,                   :foreign_key => :ems_id, :dependent => :destroy
+
+    def self.raw_api_endpoint(hostname, port)
```
So, class method should not be in the `included` section.  Create a `module ClassMethods` and define this as an instance method in there.  See [here](http://api.rubyonrails.org/classes/ActiveSupport/Concern.html).

Fryguy commented on vmdb/app/models/mixins/container_provider_mixin.rb:
```diff
@@ -0,0 +1,42 @@
+    end
+
+    # UI methods for determining availability of fields
+    def supports_port?
```
These instance methods should live outside the `included` and just be normal module instance methods at the `ContainerProviderMixin` level

Fryguy commented on vmdb/app/models/mixins/container_provider_mixin.rb:
```diff
@@ -0,0 +1,42 @@
+    def authentication_status_ok?(_type = nil)
+      # TODO: support real authentication using certificates
+      true
+    end
```
We should coordinate #2156.  I think we should merge that one first, and then this one should rebase against it.  Note that you might need verification of the OpenShift endpoint that's different from kubernetes, or perhaps checks both.

Fryguy commented on vmdb/db/migrate/20150414085027_create_container_routes.rb:
```diff
@@ -0,0 +1,21 @@
+      t.string     :resource_version
+      t.string     :namespace
+      t.string     :host
+      t.string     :service_name
```
I noticed you mentioned this should be service_id...is the intent to do that in this PR?

abonas replied:
> I noticed you mentioned this should be service_id...is the intent to do that in this PR?

No, I'll do it separately

Fryguy commented on vmdb/db/migrate/20150414085027_create_container_routes.rb:
```diff
@@ -0,0 +1,21 @@
+      t.string     :namespace
+      t.string     :host
+      t.string     :service_name
+      t.string     :path
```
`path` is an interesting method name.  If you use RelationshipMixin in the future you might get a collision, so we should discuss this column name.

abonas replied:
> path is an interesting method name. If you use RelationshipMixin in the future you might get a collision, so we should discuss this column name.

renaming suggestions are welcome. the meaning of this property on openshift side is the relative path on the server that is specified in the host field on which you commented below.
so if host = "www.example.com", path can be /services so later openshift will parse them as www.example.com/services

Fryguy commented on vmdb/db/migrate/20150414085027_create_container_routes.rb:
```diff
@@ -0,0 +1,21 @@
+      t.timestamp  :creation_timestamp
+      t.string     :resource_version
+      t.string     :namespace
+      t.string     :host
```
host is probably a bad column name.  I'm concerned it will be confused with a reference to the host table at the model level

abonas replied:
> host is probably a bad column name. I'm concerned it will be confused with a reference to the host table at the model level

I'll change it to host name. (host is the original name of that field in openshift...)

Fryguy commented:
This is an awesome PR! :clap:  I think you're really close.

abonas commented:
@Fryguy thanks for this review! will start to address comments shortly.
the main issue in this PR as I mentioned in the first comment is that there's a lacking functionality here of pulling inventory from k8s in addition to openshift. 
So what would be the appropriate way and location in code to do it? (that is, in addition to the connect.all_entities that goes to openshift_client gem, also to do the same with call to kubeclient)

Fryguy commented:
Since you are subclassing the parser, can't you just leverage the `super` methods?

abonas commented:
> Since you are subclassing the parser, can't you just leverage the super methods?
@Fryguy 
for parsing, yes, but if the data is not collected so how does that help?

Fryguy commented:
Does't the parser collect the data?  I'm not sure I understand the difference.

abonas commented:
> Does't the parser collect the data? I'm not sure I understand the difference.

Yes it does. Currently it does "all_entities = ems.with_provider_connection(&:all_entities)"
That part collects only openshift entities via the openshift_gem.
In addition, there's a need in a separate call to kubeclient to get "all_entities" from there too.
where/how should it be done?

Fryguy commented:
So, yeah, with_provider_connection, under the covers, uses the .connect method, which in turn calls raw_connect.  To deal with "services" like in the Amazon provider, you can pass something `:service => "SQS"` to get connections for alternative services, and the connect method should deal with that.  So, if you follow that pattern, you can do something like `ems.with_provider_connection(:service => :kubernetes, &:all_entities)`.

tl; dr: follow the amazon pattern and I think you'll get what you want.

Fryguy commented:
Now that the Kubernetes VCR tests are in, can we do the same for OpenShift?

abonas commented:
> Now that the Kubernetes VCR tests are in, can we do the same for OpenShift?

of course, in a separate PR. this one has grown too big.

abonas commented:
@h-kataria , @dclarizio  pls review the UI port autofill handling part.
@Fryguy , I think I addressed all your previous comments, please review. 
iirc, one open issue is the path field on container route, afaik it's not critical for now, but I'm open for alternative naming suggestions.

abonas commented:
@Fryguy , this build has failed for unrelated to the patch reasons. can you restart it please?

abonas commented:
@Fryguy , this build has failed for unrelated to the patch reasons. can you restart it please?
(and please re-review while at it, if possible :) )

simon3z commented on vmdb/app/models/ems_openshift.rb:
```diff
@@ -0,0 +1,39 @@
+    osclient
+  end
+
+  def self.kubernetes_connect(hostname, port)
```
This is the same code we have here:

https://github.com/ManageIQ/manageiq/blob/104f784/vmdb/app/models/ems_kubernetes.rb#L17-L24

we should unify it (we shouldn't remember to update two different places in case kubernetes connection changes).

abonas replied:
> This is the same code we have here:
> 
> https://github.com/ManageIQ/manageiq/blob/104f784/vmdb/app/models/ems_kubernetes.rb#L17-L24
> 
> we should unify it (we shouldn't remember to update two different places in case kubernetes connection changes).

yes, I'll unify them.

simon3z commented on vmdb/db/migrate/20150414094834_create_container_projects.rb:
```diff
@@ -0,0 +1,18 @@
+class CreateContainerProjects < ActiveRecord::Migration
+  def up
+    create_table :container_projects do |t|
```
How these are different from namespaces?

I think that even when introducing OpenShift we shouldn't fall into the mistake to represent 1:1 its entities. For example projects seems to be essentially namespaces.

I think we should prioritize the namespace collection to begin with and then use projects to augment the namespace information (if there's any valuable information to add).

abonas replied:
> How these are different from namespaces?
> 
> I think that even when introducing OpenShift we shouldn't fall into the mistake to represent 1:1 its entities. For example projects seems to be essentially namespaces.
> 
> I think we should prioritize the namespace collection to begin with and then use projects to augment the namespace information (if there's any valuable information to add).

namespace collection effort is being developed in parallel.

Here's my take on namespace vs project - currently they both should be collected and persisted.
Projects and namespaces are indeed quite similar to each other. they also share (imo - unfortunately, they are different object types after all) the same uid.
Currently, projects have additional attributes and annotations that namespaces don't have. While both k8s and ose are not stable, it's hard to predict how far those objects will grow from each other.
Also, each project is currently backed by a namespace, but it's not guaranteed afaik that each namespace has a parallel project, even in an openshift env.
Please also keep in mind, that manageIQ UI has static menus for all entities that doesn't change based on each provider selection, and the entities menu is displayed for both k8s and ose providers together, as they are both considered a container based provider.
Hence, while we can represent/navigate from each provider to its relevant entities (from k8s to ns, from ose to project), having only projects in the menu is misleading. (especially if a user has only k8s environments and doesn't even know what a project is)
Once k8s and openshift stabilize behavior and api, we can revisit the design.

Adding @derekwaynecarr for additional insights.

simon3z replied:
We have to keep in mind that the idea was to abstract these concepts, and not to create 1:1 tables with the current solution we're using (for example `container_group` is used for kubernetes pods and amazon tasks).

How data is displayed has nothing to do with how it's stored. If we decide that for clarity we need both "Projects" and "Namespaces" in the menu (I don't think we need both, just as we don't mention other entities as "Pods") then we can have them regardless of how many tables we use and how we deal with entities.

I am fine with maintaining this table and name `container_projects` and use it for namespaces as well.

abonas commented:
@Fryguy, pls get this another round of review.
I rebased and fixed @simon3z comment regarding using the same method for k8s conn in both providers.
Getting this PR in will help since there are several more patches depedent on it (VCR, UI, etc.)
Thanks

abonas commented:
@Fryguy / @blomquisg can this be merged please? 
It delays a lot of other dependent patches and it also must be rebased practically every day.

jrafanie commented on vmdb/app/controllers/ems_common.rb:
```diff
@@ -702,6 +702,10 @@ def get_form_vars
       if ["openstack", "openstack_infra"].include?(params[:server_emstype])
         @edit[:new][:port] = @ems.port ? @ems.port : 5000
+      elsif params[:server_emstype] == EmsKubernetes.ems_type
+        @edit[:new][:port] = @ems.port ?  @ems.port : EmsKubernetes.new.port
```
@abonas I think this should be a constant, EmsKubernetes::DEFAULT_PORT and have dafault_value_for use that constant.  I don't see the need to instantiate an object here.

This can be fixed in a followup PR.

abonas replied:
> @abonas I think this should be a constant, EmsKubernetes::DEFAULT_PORT and have dafault_value_for use that constant. I don't see the need to instantiate an object here.
> 
> This can be fixed in a followup PR.

@jrafanie , ack. lets get this merged and I'll fix it separately. (I filed an issue so it won't be lost)

jrafanie commented:
LGTM.

jrafanie commented:
cc @Fryguy, please do final review

abonas commented:
it needs to be rebased against https://github.com/ManageIQ/manageiq/pull/2891
and then add the additional openshift class to automation.

abonas commented:
rebased and contains the additional openshift class in automation.

abonas commented:
@Fryguy , this is all yours :)
(is travis stuck again?)

jrafanie commented:
Restarted @abonas

abonas commented:
@Fryguy @blomquisg @jrafanie anything prevents this from being merged?
